### ISSUE TITLE
SIE WAITSTATE Assist broken

### ISSUE THREAD

**Issue by Fish-Git** [Fish-Git/hyperion#91]
(this is a copy of a problem in [herules-390/issues/238](https://github.com/hercules-390/hyperion/issues/238) which also exists in SDL Hyperion)

Ivan Warren writes:

> When a z/Arch hypervisor dedicates a processor to a guest and WAITSTATE Assist is enabled (via `FEATURE_WAITSTATE_ASSIST` in feats390.h and/or feat900.h), an error occurs in "sie.c" near line 1081 with rc=22 (EINVAL) on `wait_timed_condition()` (mapping to `pthread_cond_timedwait()` on POSIX systems).
> 
> This leads to the call to always wait for the timeout to occur (regardless of any interrupt) when a SIE guest enters a wait state and the processor is dedicated to the guest, and hence severly reduces performance.
> 
> Disabling the feature in feat390.h and feat900.h alleviates the issue, but as a result forces the hypervisor to resolve guest wait states for dedicated processors when it really shouldn't be necessary.

**Comment by Fish-Git** [comment]
What is "Wait State Assist" and how is it supposed to behave?

Looking at the code it appears it is waiting for any interrupt to occur (`regs->intcond`), but only if it occurs within a specific time limit.

My first question (besides what exactly is "Wait State Assist") is, why is there a time limit at all? (and what should the time limit be? Does the documentation for "Wait State Assist" define that value?)

Second, RC=22 (EINVAL) is documented as:

> The _abstime_ argument specified a nanosecond value less than zero or greater than or equal to 1000 million.

which means we are simply not calculating our deadline (timeout) value properly (if we should even have one; see immediately preceding question), which seems to be an easy enough fix.

The `host_tod()` function returns an emulated (z/Arch) "TOD clock" value, which is _not_ the same thing as "the current Hercules host (Windows/Linux) time-of-day specified as number of microseconds since the Hercules host (Windows/Linux) clock epoch", which seems to be the way the problematic code is written:

```
    typedef U64  TOD;  // one microsecond = Bit 51
    ...
    U64 now = host_tod();
    waittime.tv_sec = now / 1000000;
    waittime.tv_nsec = ((now % 1000000) + 3333) * 1000;
    ...
    timed_wait_condition( ... &waittime );
```

I believe `now` simply needs to be shifted right by 13 bits (64-51) when used in the above `waittime` calculations. I believe that alone _should_ fix the problem. Yes?

Or again, am I missing something?

Feedback appreciated!

**Comment by Fish-Git** [comment]
> I believe that alone should fix the problem.

Actually it won't. Some additional code is need to account for overflow since 3333 microseconds is being added before being multiplied by 1000 (to convert to nanoseconds). I.e. the fix _might_ be as easy as:

```
    waittime.tv_sec  =   (now >> (64-51)) / ONE_MILLION;
    waittime.tv_nsec = (((now >> (64-51)) % ONE_MILLION) + 3333) * 1000;
    waittime.tv_sec  += waittime.tv_nsec /  ONE_BILLION;
    waittime.tv_nsec %=                     ONE_BILLION;
```

What do you think?

**Comment by Fish-Git** [comment]
> This looks correct to me.

Thanks, Peter.

However, I'd **_still_** like to know (if anyone knows!) what exactly _is_ "Wait State Assist", and how is it supposed to behave! Is it documented anywhere?

**Comment by Fish-Git** [comment]
Fixed by commit bcc28c8ead8650df6cf80270181fabf0be9c7643
